c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -208,18 +208,31 @@ impl<'a, 'tcx> UnlowerVisitor<'a, 'tcx> {
 
         // Most exprs end with an assignment, storing the result into a temporary.
         match ex.kind {
-            hir::ExprKind::Assign(pl, rv, _span) => {
+            hir::ExprKind::Assign(pl, _rv, _span) => {
                 // For `Assign`, we expect the assignment to be the whole thing.
-                let (loc, _mir_pl, _mir_rv) = match self.get_sole_assign(&locs) {
+                let (loc, _mir_pl, mir_rv) = match self.get_sole_assign(&locs) {
                     Some(x) => x,
                     None => {
                         warn("expected exactly one StatementKind::Assign");
                         return;
                     }
                 };
+                let desc = match mir_rv {
+                    mir::Rvalue::Use(op) => op
+                        .place()
+                        .map(|pl| {
+                            if is_var(pl) && self.mir.local_kind(pl.local) == mir::LocalKind::Temp {
+                                MirOriginDesc::LoadFromTemp
+                            } else {
+                                MirOriginDesc::Expr
+                            }
+                        })
+                        .unwrap_or(MirOriginDesc::Expr),
+                    _ => MirOriginDesc::Expr,
+                };
                 self.record(loc, &[], ex);
                 self.record(loc, &[SubLoc::Dest], pl);
-                self.record(loc, &[SubLoc::Rvalue], rv);
+                self.record_desc(loc, &[SubLoc::Rvalue], ex, desc);
```

Comment:
What's the reason for the change from `rv` to `ex` here?  This seems incorrect.  Using `ex` here says that MIR rewrites attached to the `Rvalue` of this MIR statement should be applied to the whole assignment.  So, for example, if there's a cast on the `Rvalue`, then I think it would get rewritten as `(x = y) as T;` instead of `x = y as T;`.

Reply:
I think you're right that it's incorrect. Does b56abb41 address the issue?

Reply:
Yes, that looks better.